dashboard/src/components/ui/CodeEditor.vue

```diff
@@ -0,0 +1,56 @@
+<template>
+  <Codemirror
+    v-model="code"
+    :style="{ height: '300px' }"
+    :autofocus="true"
+    :indent-with-tab="true"
+    :tab-size="2"
+    :extensions="extensions"
+    @ready="handleReady"
+  />
+</template>
+<script setup>
+import { ref, computed, shallowRef } from 'vue'
+import { Codemirror } from 'vue-codemirror'
+import { javascript } from '@codemirror/lang-javascript'
+import { markdown } from '@codemirror/lang-markdown'
+import { html } from '@codemirror/lang-html'
+import { json } from '@codemirror/lang-json'
+
+const props = defineProps({
+  lang: {
+    type: String,
+    default: 'html',
+  },
+})
+
+const code = defineModel({ type: String })
+
+const languages = computed(() => {
+  if (props.lang == 'javascript' || props.lang == 'js') {
+    return javascript()
+  } else if (props.lang == 'markdown' || props.lang == 'md') {
+    return markdown()
+  } else if (props.lang == 'json') {
+    return json()
+  }
+
+  return html()
+})
+
+const extensions = ref([languages.value])
+
+const view = shallowRef()
+const handleReady = (payload) => {
+  view.value = payload.view
+}
+
+const getCodemirrorStates = () => {
+  const state = view.value.state
+  const ranges = state.selection.ranges
+  const selected = ranges.reduce((r, range) => r + range.to - range.from, 0)
+  const cursor = ranges[0].anchor
+  const length = state.doc.length
+  const lines = state.doc.lines
+}
+</script>
```

Comment:
q: not sure why we need a codeeditor.

is it to support the html email type? if so, i would prefer not being able to write emails in HTML and reducing this code instead of adding the additional dependencies and code

Reply:
HTML and Markdown both are handled through this. I think offering different message input types would be a really good UX, and i can already see people asking for that feature sometime in future even if we remove it rn